packages/manager/apps/pci-instances/src/data/hooks/instance/selectors/instances.selector.ts

```diff
@@ -0,0 +1,159 @@
+import {
+  TAddress,
+  TInstance,
+  TInstanceAction,
+  TInstanceActions,
+  TInstanceAddressType,
+  TInstanceStatus,
+  TInstanceStatusSeverity,
+} from '@/types/instance/entity.type';
+import { TActionName } from '@/types/instance/common.type';
+
+const buildInstanceStatusSeverity = (
+  status: TInstanceStatusDto,
+): TInstanceStatusSeverity => {
+  switch (status) {
+    case 'BUILDING':
+    case 'REBOOT':
+    case 'REBUILD':
+    case 'REVERT_RESIZE':
+    case 'SOFT_DELETED':
+    case 'VERIFY_RESIZE':
+    case 'MIGRATING':
+    case 'RESIZE':
+    case 'BUILD':
+    case 'SHUTOFF':
+    case 'RESCUE':
+    case 'SHELVED':
+    case 'SHELVED_OFFLOADED':
+    case 'RESCUING':
+    case 'UNRESCUING':
+    case 'SNAPSHOTTING':
+    case 'RESUMING':
+    case 'HARD_REBOOT':
+    case 'PASSWORD':
+    case 'PAUSED':
+      return 'warning';
+    case 'DELETED':
+    case 'ERROR':
+    case 'STOPPED':
+    case 'SUSPENDED':
+    case 'UNKNOWN':
+      return 'error';
+    case 'ACTIVE':
+    case 'RESCUED':
+    case 'RESIZED':
+      return 'success';
+    default:
+      return 'info';
+  }
+};
+
+const getInstanceStatus = (status: TInstanceStatusDto): TInstanceStatus => ({
+  state: status,
+  severity: buildInstanceStatusSeverity(status),
+});
+
+const getActionHrefByName = (
+  projectUrl: string,
+  name: TActionName,
+  { region, id }: Pick<TInstance, 'id' | 'region'>,
+): TInstanceAction['link'] => {
+  if (name === 'details') {
+    return { path: id, isExternal: false };
+  }
+
+  if (name === 'create_autobackup') {
+    return { path: `${projectUrl}/workflow/new`, isExternal: true };
+  }
+
+  if (name === 'assign_floating_ip') {
+    const searchParams = new URLSearchParams({
+      ipType: 'floating_ip',
+      region,
+      instance: id,
+    });
+
+    return {
+      path: `${projectUrl}/public-ips/order?${searchParams.toString()}`,
+      isExternal: true,
+    };
+  }
+
+  if (name === 'soft_reboot') {
+    return {
+      path: `region/${region}/instance/${id}/soft-reboot`,
+      isExternal: false,
+    };
+  }
+
+  if (name === 'hard_reboot') {
+    return {
+      path: `region/${region}/instance/${id}/hard-reboot`,
+      isExternal: false,
+    };
+  }
+
+  const actions = new Set([
+    'delete',
+    'stop',
+    'start',
+    'shelve',
+    'unshelve',
+    'reinstall',
+  ]);
+
+  if (actions.has(name)) {
+    return {
+      path: `region/${region}/instance/${id}/${name}`,
+      isExternal: false,
+    };
+  }
+
+  return { path: '', isExternal: false };
+};
+
+const mapInstanceAddresses = (instance: TInstanceDto) =>
+  instance.addresses.reduce((acc, { type, ...rest }) => {
+    const foundAddresses = acc.get(type);
+    const ipAlreadyExists = !!foundAddresses?.find(({ ip }) => ip === rest.ip);
+    if (foundAddresses) {
+      if (ipAlreadyExists) return acc.set(type, [...foundAddresses]);
+      return acc.set(type, [...foundAddresses, rest]);
+    }
+    return acc.set(type, [rest]);
```

Comment:
```suggestion
    if (acc.has(type)) {
      const foundAddresses = acc.get(type);
      const ipAlreadyExists = !!foundAddresses.find(({ ip }) => ip === rest.ip);
      if (!ipAlreadyExists) foundAddresses.push(rest);
      return acc;
    }
    return acc.set(type, [rest]);
```
Also it seems to be weird to use a HashMap for adresses types but not for adresses ip

Reply:
I'm not a big fan using `push` fonction as it mutates the original array, I'll prefer keep it immutable :)

Reply:
Agree that your solution is OK, thx :)